### PR TITLE
DM-39390: Add edge flag to trailed sources

### PR DESCRIPTION
**Pull request by bsmartradio**
Add a flag in NaivePlugin.py to check if the trailed source extends off the image or if either the beginning or end of the trail contains edge pixels. If any of these cases are true, set `ext_trailedSources_Naive_flag_edge` to True. Additionally, added several unit tests to check the new flags.

### CODE REVIEW

**Comment by bsmartradio** [comment]
Jenkins run: https://rubin-ci.slac.stanford.edu/blue/organizations/jenkins/stack-os-matrix/detail/stack-os-matrix/301/pipeline

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 1]
```diff
@@ -124,6 +124,7 @@ def __init__(self, config, name, schema, metadata, logName=None):
         self.NO_CONVERGE = flagDefs.add("flag_noConverge", "The root finder did not converge")
         self.NO_SIGMA = flagDefs.add("flag_noSigma", "No PSF width (sigma)")
         self.SAFE_CENTROID = flagDefs.add("flag_safeCentroid", "Fell back to safe centroid extractor")
+        self.EDGE = flagDefs.add("flag_edge", "trail contains edge pixels or extends off chip")
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py]:
```suggestion
        self.EDGE = flagDefs.add("flag_edge", "Trail contains edge pixels or extends off chip")
```

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 2]
```diff
@@ -191,6 +192,20 @@ def measure(self, measRecord, exposure):
         x1 = xc + dydtheta
         y1 = yc + dxdtheta
 
+        if not (0 <= x0 <= exposure.getBBox().endX and 0 <= x1 <= exposure.getBBox().endX
+                and 0 <= y0 <= exposure.getBBox().endY and 0 <= y1 <= exposure.getBBox().endY):
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: Are you certain the edge points are appropriately 0 and not, e.g., `exposure.getBBox().startX` ?

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 3]
```diff
@@ -191,6 +192,20 @@ def measure(self, measRecord, exposure):
...
+
+            self.flagHandler.setValue(measRecord, self.EDGE.number, True)
+
+        else:
+
+            if exposure.mask[np.round(x0), np.round(y0)] and exposure.mask[np.round(x1), np.round(y1)]:
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: An inline comment or two here would enhance clarity about what is happening

File: tests/test_trailedEdgeSources.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+import numpy as np
+import unittest
+import lsst.utils.tests
+import lsst.meas.extensions.trailedSources
+from lsst.meas.base.tests import AlgorithmTestCase
+from lsst.utils.tests import classParameters
+
+# Trailed-source length, angle, and centroid, one on the edge, one not.
+Ls = np.array([5, 5])
```
Comment (mrawls) [inline tests/test_trailedEdgeSources.py]: Suggest more descriptive variable names so you can just say "one trailed source is on the edge and one is not." e.g., `trail_lengths`, `trail_angles`, `trail_x_centroids`, `trail_y_centroids`. If you really want to keep most of these and only change one, `Ls` is by far my least favorite 😄

File: tests/test_trailedEdgeSources.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+    def __init__(self, bbox, threshold=10.0, exposure=None, **kwds):
+
+        super().__init__(bbox, threshold, exposure, **kwds)
+
+    def addTrailedSource(self, trail):
+        """Add a trailed source to the simulation.
+        'Re-implemented' version of
+        `lsst.meas.base.tests.TestDataset.addSource`. Numerically integrates a
+        Gaussian PSF over a line to obtain am image of a trailed source.
```
Comment (mrawls) [inline tests/test_trailedEdgeSources.py]:
```suggestion
        Gaussian PSF over a line to obtain an image of a trailed source.
```

File: tests/test_trailedEdgeSources.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        edge_pixel_set = np.array([(exposure.mask[x0, y0] & exposure.mask.getPlaneBitMask(
+            'EDGE') != 0), (exposure.mask[x1, y1] & exposure.mask.getPlaneBitMask('EDGE') != 0)])
+
+        # Checks trailed which crosses into edge pixels
+        if record['truth_x'] == 100:
+
+            self.assertTrue(edge_pixel_set.any())
+
+            # Make sure measurement edge flag is set, but Naive_flag is not.
+            # A failed trailed source measurement can still have the edge flag
```
Comment (mrawls) [inline tests/test_trailedEdgeSources.py]: What would having the new edge flag set mean in the situation of a failed measurement? Would it be worthwhile to add a null test, i.e., one where a measurement fails and the flag is not errantly set as a result of the failure? Part of this question is coming from an uncertainty of what failed measurement type situations would ever cause `Naive_flag` to be set...

File: tests/test_trailedEdgeSources.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    """A dataset for testing trailed source measurements.
+    Given a `TrailedSource`, construct a record of the true values and an
+    Exposure.
+    """
+
+    def __init__(self, bbox, threshold=10.0, exposure=None, **kwds):
+
+        super().__init__(bbox, threshold, exposure, **kwds)
+
+    def addTrailedSource(self, trail):
```
Comment (mrawls) [inline tests/test_trailedEdgeSources.py]: This is a clever way of generating fake trails! Do you have any images showing that it is doing what you think it is? I worry it is easy to have a geometry error or be off by a factor of 2 or something "silly" without actually seeing any fake trails and their positions.

**Review by mrawls — COMMENTED** [review]
This appears to do what it says on the tin! My comments are pretty minor and I don't see anything blocking merging. Thank you for writing a test for this new functionality. Please do squash your commits appropriately and write a brief description in the PR.

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 1 above)
Comment (bsmartradio) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: Fixed

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 2 above)
Comment (bsmartradio) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: OIh, you are right, thanks for the catch. I swapped it this way as a test and missed swapping it back. I'll make sure to run Jenkins again to verify it doesn't change anything.

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 3 above)
Comment (bsmartradio) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: Done

File: tests/test_trailedEdgeSources.py (hunk 5 above)
Comment (bsmartradio) [reply tests/test_trailedEdgeSources.py]: Fixed

**Comment by bsmartradio** [comment]
Latest Jenkins run: https://rubin-ci.slac.stanford.edu/blue/organizations/jenkins/stack-os-matrix/detail/stack-os-matrix/365/pipeline

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 8]
```diff
@@ -191,6 +192,24 @@ def measure(self, measRecord, exposure):
         x1 = xc + dydtheta
         y1 = yc + dxdtheta
 
+        if not (exposure.getBBox().beginX <= x0 <= exposure.getBBox().endX
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py:196]: Add one-line comment, e.g., `# Check whether trail extends into the edge of the exposure`

File: tests/test_trailedEdgeSources.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+    """
+
+    def __init__(self, bbox, threshold=10.0, exposure=None, **kwds):
+
+        super().__init__(bbox, threshold, exposure, **kwds)
+
+    def addTrailedSource(self, trail, edge=True):
+        """Add a trailed source to the simulation.
+
+         Re-implemented version of
```
Comment (mrawls) [inline tests/test_trailedEdgeSources.py]:
```suggestion
        Re-implemented version of
```

Comment (mrawls) [reply tests/test_trailedEdgeSources.py]: (I think that makes the block comment indentation consistent?)

File: tests/test_trailedEdgeSources.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+
+# Following from test_trailedSources
+@classParameters(length=length, theta=angle, xc=x_coord, yc=y_coord)
+class TrailedEdgeSourcesTestCase(AlgorithmTestCase, lsst.utils.tests.TestCase):
+    """ Test if ext_trailedSources_Naive_flag_edge is set correctly.
+
+    Given a `TrailedSource`, test if the edge flag is set correctly in the
+    source catalog after the
+    'lsst.meas.extensions.trailedSources.Naive.Plugin.makeTrailedSourceMeasurementTask'
```
Comment (mrawls) [inline tests/test_trailedEdgeSources.py]: You want backticks here, I believe, not single quotes

File: tests/test_trailedEdgeSources.py [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+
+        # Check that x0, y0 or x1, y1 is flagged as an edge pixel
+        x1 = int(record['ext_trailedSources_Naive_x1'])
+        y1 = int(record['ext_trailedSources_Naive_y1'])
+        x0 = int(record['ext_trailedSources_Naive_x0'])
+        y0 = int(record['ext_trailedSources_Naive_y0'])
+
+        # Test Case with no edge pixels
+        if record['truth_x'] == 100:
+
```
Comment (mrawls) [inline tests/test_trailedEdgeSources.py]: Suggest removing blank lines after the if / elif / else lines

**Review by mrawls — APPROVED** [review]
Thanks for taking the time to make sure this did what it was supposed to do! 😄

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 8 above)
Comment (bsmartradio) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py:196]: Added

File: tests/test_trailedEdgeSources.py (hunk 9 above)
Comment (bsmartradio) [reply tests/test_trailedEdgeSources.py]: Fixed

File: tests/test_trailedEdgeSources.py (hunk 10 above)
Comment (bsmartradio) [reply tests/test_trailedEdgeSources.py]: Fixed

File: tests/test_trailedEdgeSources.py (hunk 11 above)
Comment (bsmartradio) [reply tests/test_trailedEdgeSources.py]: Removed

**Pull request merged**